base/strings/search.jl

```diff
@@ -212,44 +206,50 @@ julia> rsearch("aaabbb","b")
 6:6
 ```
 """
-function rsearch(s::AbstractString, c::Chars, i::Integer)
-    e = endof(s)
-    j = search(RevString(s), c, e-i+1)
-    j == 0 && return 0
-    e-j+1
+function rsearch(s::AbstractString, c::Chars, i::Integer=start(s))
+    @boundscheck checkbounds(s, i)
+    isempty(c) && return i
+    j = i
+    while j > 0
```

Comment:
`@inbounds` would be a good idea since you've checked the index. Same in similar functions.

In general it seems to me that the iteration and indexing API isn't optimized for backward iteration: for example, `is_valid_continuation` will be called first by `prevind`, and then again by `next`. But I'm not sure it makes a big difference, and for now I'd say it's OK.